writePSQLFile: init

MatthewCroughan opened the pull request:
Adds a function called writePSQLFile which validates the input SQL text, by running it through postgresql via the postgresqlTestHook created at [nix.camp 2022](https://nix.camp)

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

sarcasticadmin commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -250,6 +250,39 @@ rec {
+   *    '';
+   * }
+  */
+  writeSQL = name: text:
```
was the name supposed to be `writePSQL`?

MatthewCroughan replied:
Yup

sarcasticadmin reviewed (approved):
@MatthewCroughan looks awesome!

FRidh commented:
@ofborg eval

gavinrogers commented:
This seems like an elegant refinement.notich is needed in the way of testing, but if i understand right, ot could break a lot of legacy sqp (this is a meta situation).

Could someone at the nix.camp fire up tree and sbuild a pile of (if not all amd any which contain sql , either in the derivation or in the installer (tribler used to do this)) derivations with this check added

@zimbatm this seems like up your alley, could we trouble you for a thumbs up or amy other feedback? I cam give you a lush ride to/from the camp for your helpful critixism)

zimbatm reviewed (changes requested):
Are there plans to use that helper function in nixpkgs itself?

Without actual usage, these types of functions tend to rot, so it's best to keep them out of nixpkgs. Otherwise, let's also add some usage of it to this PR.

Technically it looks valid

MatthewCroughan commented:
@zimbatm I was making a NixOS VM Test with https://github.com/NixOS/nixpkgs/pull/191201 and it would have been nice to use writePSQL to write the sql files for the initialScript in that test, for example.

zimbatm commented:
Are there any other use cases you can think of?

I'm the grumpy old that doesn't want children to have fun :)

Usually, the best is to start with the use cases, and then extract the common functions. That way, we don't end up with a super large API surface that is unmaintained.

SuperSandro2000 commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -250,6 +250,26 @@ rec {
+  writePSQL = name: text:
+    writeTextFile {
+      name = "${name}.sql";
+      inherit text;
+      checkPhase =
+        ''
+          runHook preCheck
+          ${pgformatter}/bin/pg_format "$target"
+          runHook postCheck
+        '';
```
```suggestion
      inherit text;
      checkPhase = ''
        runHook preCheck
        ${buildPackages.pgformatter}/bin/pg_format "$target"
        runHook postCheck
      '';
```
Don't we require nativeBuildInputs (which would require runCommand) to support cross?

MatthewCroughan commented:
@zimbatm @SuperSandro2000 I've refactored this function in order to use `psql` and the `postgresqlTestHook` instead of some perl script (pgformatter). Please review this instead :D

Profpatsch commented:
Does postgresql not have a static check mode, where the file is just checked for whether the engine would accept it? I feel like loading it into a temporary database might not work for a lot of files (e.g. you can’t run any `UPDATE`s without already having a table set up).

MatthewCroughan commented:
@Profpatsch That is the intention of this PR, and if you know of such an option, that would be good to use.

MatthewCroughan committed:
build-support/writeTextFile: pass doCheck and checkInputs

This allows us to easily write functions which use writeTextFile to
validate or sanitize input text

MatthewCroughan committed:
writePSQLFile: init

Adds a function called writePSQLFile which validates the input SQL text,
by running it through postgresql via the postgresqlTestHook

MatthewCroughan commented:
> Are there plans to use that helper function in nixpkgs itself?
> 
> Without actual usage, these types of functions tend to rot, so it's best to keep them out of nixpkgs. Otherwise, let's also add some usage of it to this PR.
> 
> Technically it looks valid

There cannot be actual usage until it is merged into nixpkgs, but I do understand that a lot of the writers are underutilized.
